estimator/ntru_primal.py

```diff
@@ -0,0 +1,503 @@
+
+        remaining_proba = RR(1.)
+        average_beta = RR(0.)
+        total_DSD_prob = RR(0.)
+        DSD_prob = RR(0.)
+        prob_pos_total = np.zeros(2*params.n, dtype='double')
+
+        for beta in range(2, params.n):
+            tours = floor(params.n**2 / beta**2)+3
+
+            DSD_prob, DSD_prob_pos = self.prob_dsd(beta, params, red_shape_model, m=m,
+                                                   red_cost_model=red_cost_model, log_level=log_level)
+
+            if DSD_prob > 10e-8:
+                for t in range(tours):
+                    for i in range(2*params.n):
+                        prob_pos = DSD_prob_pos[i]
+                        average_beta += RR(beta) * remaining_proba * prob_pos
+                        prob_pos_total[i] += remaining_proba * prob_pos
+                        total_DSD_prob += remaining_proba * prob_pos
+                        remaining_proba *= (1.-prob_pos)
+
+                Logging.log("dsd", log_level+1, "β= %d,\t pr=%.4e, \t rem-pr=%.4e"%(beta, DSD_prob, remaining_proba))
+            if remaining_proba < 0.001:
+                average_beta += beta * remaining_proba
+                break
+
+        if not average_beta:
+            average_beta = m
+            predicate = False
+
+        else:
+            predicate = True
+
+        cost = costf(red_cost_model, RR(average_beta), m, predicate=predicate)
+        cost["tag"] = "dsd"
+        cost["problem"] = params
+        return cost.sanity_check()
+
+    __name__ = "primal_dsd"
+
+
+primal_dsd = PrimalDSD()
+
+
+class NTRUPrimalUSVP(PrimalUSVP):
+
+    @staticmethod
+    def _solve_for_d(params, m, beta, tau, xi):
+        """
+        Find smallest d ∈ [n,m] to satisfy uSVP condition.
+
+        If no such d exists, return the upper bound m.
+        :tau: ignored due to NTRU being homogeneous
+        """
+        # Find the smallest d ∈ [n,m] s.t. a*d^2 + b*d + c >= 0
+
+        delta = deltaf(beta)
+        a = -log(delta)
+        C = log(params.Xe.stddev**2 * (beta - 1)) / 2.0
+        b = log(delta) * (2 * beta - 1) + log(params.q) - C
+        c = params.n * log(xi) - (params.n + 1) * log(params.q)
+        n = params.n
+        if a * n * n + b * n + c >= 0:  # trivial case
+            return n
+
+        # solve for ad^2 + bd + c == 0
+        disc = b * b - 4 * a * c  # the discriminant
+        if disc < 0:  # no solution, return m
+            return m
+
+        # compute the two solutions
+        d1 = (-b + sqrt(disc)) / (2 * a)
+        d2 = (-b - sqrt(disc)) / (2 * a)
+        if a > 0:  # the only possible solution is ceiling(d2)
+            return min(m, ceil(d2))
+
+        # the case a<=0:
+        # if n is to the left of d1 then the first solution is ceil(d1)
+        if n <= d1:
+            return min(m, ceil(d1))
+
+        # otherwise, n must be larger than d2 (since an^2+bn+c<0) so no solution
+        return m
+
+    @staticmethod
+    @cached_function
+    def cost_gsa(
```

Comment:
Can we re-use existing code?

Reply:
Unfortunately, no. Needed to remove references to Tau to account for the fact that NTRU is homogeneous.

Reply:
We could generalise the underlying code, though? At least this function shouldn't take tau or throw an error when tau is provided?

Reply:
It's an overload, so convention dictates it has the same method signature. All of these cost functions set Tau automatically if Tau is None. But in simulator.py Tau=None is the convention for homogeneous systems. I wanted to take the route of adding new code, rather than modifying existing code. I can certainly generalize these functions.